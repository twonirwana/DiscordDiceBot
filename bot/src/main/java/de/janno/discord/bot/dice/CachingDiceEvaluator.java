package de.janno.discord.bot.dice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.ExpressionException;
import de.janno.evaluator.dice.Roller;
import de.janno.evaluator.dice.random.NumberSupplier;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.Tags;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.ExecutionException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

@Slf4j
public class CachingDiceEvaluator {

    private final LoadingCache<String, RollerOrError> diceRollerCache;
    private final DiceEvaluator diceEvaluator;

    public CachingDiceEvaluator(NumberSupplier numberSupplier, int maxNumberOfDice, int cacheSize) {
        diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
        diceRollerCache = CacheBuilder.newBuilder()
                .maximumSize(cacheSize)
                .recordStats()
                .build(new CacheLoader<>() {
                    @Override
                    public @NonNull RollerOrError load(@NonNull String expression) {
                        try {
                            log.debug("create roller for: {}", expression);
                            Roller roller = diceEvaluator.buildRollSupplier(expression);
                            roller.roll();
                            return new RollerOrError(expression, roller, true, null);
                        } catch (ExpressionException | ArithmeticException e) {
                            return new RollerOrError(expression, null, false, e.getMessage());
                        }
                    }
                });
        Gauge.builder("diceEvaluator.cache", diceRollerCache::size).tags(Tags.of("stats", "size")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().requestCount()).tags(Tags.of("stats", "requests")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().hitCount()).tags(Tags.of("stats", "hit")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().missCount()).tags(Tags.of("stats", "miss")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().totalLoadTime()).tags(Tags.of("stats", "totalLoadTime")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().averageLoadPenalty()).tags(Tags.of("stats", "averageLoadTime")).register(globalRegistry);
    }

    public @NonNull RollerOrError get(@NonNull String expression) {
        try {
            return diceRollerCache.get(expression);
        } catch (ExecutionException e) {
            throw new IllegalStateException(e);
        }
    }

    @VisibleForTesting
    public long getCacheSize() {
        return diceRollerCache.size();
    }
}
