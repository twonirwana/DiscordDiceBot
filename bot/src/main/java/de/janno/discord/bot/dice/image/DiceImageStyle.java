package de.janno.discord.bot.dice.image;

import de.janno.discord.bot.dice.image.provider.*;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public enum DiceImageStyle {
    none(new NoneProvider()),
    polyhedral_3d(new PolyhedralFileImageProvider("polyhedral_3d", List.of("red_and_white"), "red_and_white")),
    polyhedral_alies_v2(new PolyhedralFileImageProvider(
            "polyhedral_alies_v2",
            List.of("blue_and_silver", "black_and_gold", "blue_and_gold", "green_and_gold", "orange_and_silver", "red_and_gold", "purple_and_silver"),
            "blue_and_silver")),
    polyhedral_knots(new PolyhedralFileImageProvider("polyhedral_knots", List.of("purple_dark", "purple_white", "blue"), "blue")),
    polyhedral_RdD(new PolyhedralFileImageProvider("polyhedral_RdD", List.of("default", "special"), "default", Map.of(
            "default", List.of(4, 6, 7, 8, 10, 12, 20, 100),
            "special", List.of(8, 12)))),
    fate(new FateBasic()),
    d6_dots(new D6Dotted()),
    polyhedral_2d(new PolyhedralSvgWithColor()),
    polyhedral_alies_v1(new PolyhedralFileImageProvider("polyhedral_alies_v1", List.of("black_and_gold"), "black_and_gold"));

    private final ImageProvider imageProvider;

    DiceImageStyle(ImageProvider imageProvider) {
        this.imageProvider = imageProvider;
    }

    public static boolean isValidStyle(String name) {
        return Arrays.stream(values()).map(Enum::name).collect(Collectors.toSet()).contains(name);
    }

    public static String combineStyleAndColorName(DiceImageStyle style, String color) {
        return "%s_%s".formatted(style.name(), color);
    }

    public ImageProvider getImageProvider() {
        return imageProvider;
    }

    public String getDefaultColor() {
        return imageProvider.getDefaultColor();
    }

    public List<String> getSupportedColors() {
        return imageProvider.getSupportedColors();
    }
}
