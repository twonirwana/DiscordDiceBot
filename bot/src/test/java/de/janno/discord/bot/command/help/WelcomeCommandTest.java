package de.janno.discord.bot.command.help;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.command.*;
import de.janno.discord.bot.command.customDice.CustomDiceCommand;
import de.janno.discord.bot.command.customParameter.CustomParameterCommand;
import de.janno.discord.bot.command.sumCustomSet.SumCustomSetCommand;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.dice.image.DiceImageStyle;
import de.janno.discord.bot.dice.image.DiceStyleAndColor;
import de.janno.discord.bot.persistance.MessageConfigDTO;
import de.janno.discord.bot.persistance.PersistenceManager;
import de.janno.discord.bot.persistance.PersistenceManagerImpl;
import de.janno.discord.connector.api.ButtonEventAdaptor;
import de.janno.discord.connector.api.DiscordConnector;
import de.janno.discord.connector.api.message.ButtonDefinition;
import de.janno.discord.connector.api.message.EmbedOrMessageDefinition;
import de.janno.evaluator.dice.random.RandomNumberSupplier;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.mockito.Mockito;

import java.util.Collection;
import java.util.Locale;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SnapshotExtension.class)
class WelcomeCommandTest {

    WelcomeCommand underTest;
    private Expect expect;

    @BeforeEach
    void setup() {
        PersistenceManager persistenceManager = Mockito.mock(PersistenceManager.class);
        CachingDiceEvaluator cachingDiceEvaluator = new CachingDiceEvaluator(new RandomNumberSupplier(0), 1000, 0);
        CustomDiceCommand customDiceCommand = new CustomDiceCommand(persistenceManager, cachingDiceEvaluator);
        CustomParameterCommand customParameterCommand = new CustomParameterCommand(persistenceManager, cachingDiceEvaluator);
        SumCustomSetCommand sumCustomSetCommand = new SumCustomSetCommand(persistenceManager, cachingDiceEvaluator);
        RpgSystemCommandPreset rpgSystemCommandPreset = new RpgSystemCommandPreset(persistenceManager, customParameterCommand, customDiceCommand, sumCustomSetCommand);
        underTest = new WelcomeCommand(persistenceManager, rpgSystemCommandPreset, () -> UUID.fromString("00000000-0000-0000-0000-000000000000"));
    }

    @Test
    public void getButtonMessageWithState_fate() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("fate", StateData.empty()), 1L, 2L);
        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Please select value for **Modifier**");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("custom_parameterid100000000-0000-0000-0000-000000000000",
                        "custom_parameterid200000000-0000-0000-0000-000000000000",
                        "custom_parameterid300000000-0000-0000-0000-000000000000",
                        "custom_parameterid400000000-0000-0000-0000-000000000000",
                        "custom_parameterid500000000-0000-0000-0000-000000000000",
                        "custom_parameterid600000000-0000-0000-0000-000000000000",
                        "custom_parameterid700000000-0000-0000-0000-000000000000",
                        "custom_parameterid800000000-0000-0000-0000-000000000000",
                        "custom_parameterid900000000-0000-0000-0000-000000000000",
                        "custom_parameterid1000000000-0000-0000-0000-000000000000",
                        "custom_parameterid1100000000-0000-0000-0000-000000000000",
                        "custom_parameterid1200000000-0000-0000-0000-000000000000",
                        "custom_parameterid1300000000-0000-0000-0000-000000000000",
                        "custom_parameterid1400000000-0000-0000-0000-000000000000",
                        "custom_parameterid1500000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("-4", "-3", "-2", "-1", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10");

    }

    @Test
    public void getButtonMessageWithState_dnd5() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("dnd5", StateData.empty()), 1L, 2L);
        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Click on a button to roll the dice");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("custom_dice1_button00000000-0000-0000-0000-000000000000",
                        "custom_dice2_button00000000-0000-0000-0000-000000000000",
                        "custom_dice3_button00000000-0000-0000-0000-000000000000",
                        "custom_dice4_button00000000-0000-0000-0000-000000000000",
                        "custom_dice5_button00000000-0000-0000-0000-000000000000",
                        "custom_dice6_button00000000-0000-0000-0000-000000000000",
                        "custom_dice7_button00000000-0000-0000-0000-000000000000",
                        "custom_dice8_button00000000-0000-0000-0000-000000000000",
                        "custom_dice9_button00000000-0000-0000-0000-000000000000",
                        "custom_dice10_button00000000-0000-0000-0000-000000000000",
                        "custom_dice11_button00000000-0000-0000-0000-000000000000",
                        "custom_dice12_button00000000-0000-0000-0000-000000000000",
                        "custom_dice13_button00000000-0000-0000-0000-000000000000",
                        "custom_dice14_button00000000-0000-0000-0000-000000000000",
                        "custom_dice15_button00000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("1d4",
                        "1d6",
                        "1d8",
                        "1d10",
                        "1d12",
                        "1d20",
                        "1d100",
                        "D20 Advantage",
                        "D20 Disadvantage",
                        "2d4",
                        "2d6",
                        "2d8",
                        "2d10",
                        "2d12",
                        "2d20");
    }

    @Test
    public void getButtonMessageWithState_coin() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("coin", StateData.empty()), 1L, 2L);
        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Click on a button to roll the dice");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("custom_dice1_button00000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("Coin Toss \uD83E\uDE99");
    }

    @Test
    public void getButtonMessageWithState_nWoD() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("nWoD", StateData.empty()), 1L, 2L);

        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Please select value for **Number of Dice**");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("custom_parameterid100000000-0000-0000-0000-000000000000",
                        "custom_parameterid200000000-0000-0000-0000-000000000000",
                        "custom_parameterid300000000-0000-0000-0000-000000000000",
                        "custom_parameterid400000000-0000-0000-0000-000000000000",
                        "custom_parameterid500000000-0000-0000-0000-000000000000",
                        "custom_parameterid600000000-0000-0000-0000-000000000000",
                        "custom_parameterid700000000-0000-0000-0000-000000000000",
                        "custom_parameterid800000000-0000-0000-0000-000000000000",
                        "custom_parameterid900000000-0000-0000-0000-000000000000",
                        "custom_parameterid1000000000-0000-0000-0000-000000000000",
                        "custom_parameterid1100000000-0000-0000-0000-000000000000",
                        "custom_parameterid1200000000-0000-0000-0000-000000000000",
                        "custom_parameterid1300000000-0000-0000-0000-000000000000",
                        "custom_parameterid1400000000-0000-0000-0000-000000000000",
                        "custom_parameterid1500000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15");
    }

    @Test
    public void getButtonMessageWithState_oWoD() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("oWoD", StateData.empty()), 1L, 2L);
        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Please select value for **Number of Dice**");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("custom_parameterid100000000-0000-0000-0000-000000000000",
                        "custom_parameterid200000000-0000-0000-0000-000000000000",
                        "custom_parameterid300000000-0000-0000-0000-000000000000",
                        "custom_parameterid400000000-0000-0000-0000-000000000000",
                        "custom_parameterid500000000-0000-0000-0000-000000000000",
                        "custom_parameterid600000000-0000-0000-0000-000000000000",
                        "custom_parameterid700000000-0000-0000-0000-000000000000",
                        "custom_parameterid800000000-0000-0000-0000-000000000000",
                        "custom_parameterid900000000-0000-0000-0000-000000000000",
                        "custom_parameterid1000000000-0000-0000-0000-000000000000",
                        "custom_parameterid1100000000-0000-0000-0000-000000000000",
                        "custom_parameterid1200000000-0000-0000-0000-000000000000",
                        "custom_parameterid1300000000-0000-0000-0000-000000000000",
                        "custom_parameterid1400000000-0000-0000-0000-000000000000",
                        "custom_parameterid1500000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15");


    }

    @Test
    public void getButtonMessageWithState_Shadowrun() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("shadowrun", StateData.empty()), 1L, 2L);
        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Please select value for **number of dice**");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("custom_parameterid100000000-0000-0000-0000-000000000000",
                        "custom_parameterid200000000-0000-0000-0000-000000000000",
                        "custom_parameterid300000000-0000-0000-0000-000000000000",
                        "custom_parameterid400000000-0000-0000-0000-000000000000",
                        "custom_parameterid500000000-0000-0000-0000-000000000000",
                        "custom_parameterid600000000-0000-0000-0000-000000000000",
                        "custom_parameterid700000000-0000-0000-0000-000000000000",
                        "custom_parameterid800000000-0000-0000-0000-000000000000",
                        "custom_parameterid900000000-0000-0000-0000-000000000000",
                        "custom_parameterid1000000000-0000-0000-0000-000000000000",
                        "custom_parameterid1100000000-0000-0000-0000-000000000000",
                        "custom_parameterid1200000000-0000-0000-0000-000000000000",
                        "custom_parameterid1300000000-0000-0000-0000-000000000000",
                        "custom_parameterid1400000000-0000-0000-0000-000000000000",
                        "custom_parameterid1500000000-0000-0000-0000-000000000000",
                        "custom_parameterid1600000000-0000-0000-0000-000000000000",
                        "custom_parameterid1700000000-0000-0000-0000-000000000000",
                        "custom_parameterid1800000000-0000-0000-0000-000000000000",
                        "custom_parameterid1900000000-0000-0000-0000-000000000000",
                        "custom_parameterid2000000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20");

    }

    @Test
    public void getButtonMessageWithState_diceCalculator() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"), new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("dice_calculator", StateData.empty()), 1L, 2L);
        assertThat(res.map(EmbedOrMessageDefinition::getDescriptionOrContent))
                .contains("Click the buttons to add dice to the set and then on Roll");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly("sum_custom_set1_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set2_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set3_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set4_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set5_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set6_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set7_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set8_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set9_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set10_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set11_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set12_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set13_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set14_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_set15_button00000000-0000-0000-0000-000000000000",
                        "sum_custom_setroll00000000-0000-0000-0000-000000000000",
                        "sum_custom_setclear00000000-0000-0000-0000-000000000000",
                        "sum_custom_setback00000000-0000-0000-0000-000000000000");
        assertThat(res.map(EmbedOrMessageDefinition::getComponentRowDefinitions)
                .stream()
                .flatMap(Collection::stream)
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly("7",
                        "8",
                        "9",
                        "+",
                        "-",
                        "4",
                        "5",
                        "6",
                        "d",
                        "k",
                        "1",
                        "2",
                        "3",
                        "0",
                        "l",
                        "Roll",
                        "Clear",
                        "Back");

    }

    @Test
    public void getButtonMessageWithState_other() {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"),
                new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>("-", StateData.empty()), 1L, 2L);
        assertThat(res)
                .isEmpty();

    }

    @Test
    public void getWelcomeMessage() {
        EmbedOrMessageDefinition res = underTest.getWelcomeMessage().apply(new DiscordConnector.WelcomeRequest(123L, 456L, Locale.ENGLISH));
        assertThat(res.getDescriptionOrContent())
                .isEqualTo("""
                        Welcome to the Button Dice Bot,
                        use one of the example buttons below to start one of the RPG dice systems, use `/quickstart system` to select one of many RPG presets or use the slash command to configure your own custom dice system (see https://github.com/twonirwana/DiscordDiceBot for details or the slash command `/help`).
                        You can also use the slash command `/r` to directly roll dice with.
                        For help or feature request come to the support discord server: https://discord.gg/e43BsqKpFr""");
        assertThat(res.getComponentRowDefinitions()
                .stream()
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getId))
                .containsExactly(
                        "welcomefate00000000-0000-0000-0000-000000000000",
                        "welcomefate_image00000000-0000-0000-0000-000000000000",
                        "welcomednd500000000-0000-0000-0000-000000000000",
                        "welcomednd5_image00000000-0000-0000-0000-000000000000",
                        "welcomenWoD00000000-0000-0000-0000-000000000000",
                        "welcomeoWoD00000000-0000-0000-0000-000000000000",
                        "welcomeshadowrun00000000-0000-0000-0000-000000000000",
                        "welcomecoin00000000-0000-0000-0000-000000000000",
                        "welcomedice_calculator00000000-0000-0000-0000-000000000000");
        assertThat(res.getComponentRowDefinitions()
                .stream()
                .flatMap(s -> s.getButtonDefinitions().stream())
                .map(ButtonDefinition::getLabel))
                .containsExactly(
                        "Fate",
                        "Fate with dice images",
                        "D&D5e",
                        "D&D5e with dice images",
                        "nWoD",
                        "oWoD",
                        "Shadowrun",
                        "Coin Toss 🪙",
                        "Dice Calculator");

    }

    @ParameterizedTest
    @CsvSource({
            "fate",
            "fate_image",
            "dnd5",
            "dnd5_image",
            "nWoD",
            "oWoD",
            "shadowrun",
            "coin",
            "dice_calculator"
    })
    void createMessageDataForNewMessage(String buttonValue) {
        Optional<EmbedOrMessageDefinition> res = underTest.createNewButtonMessageWithState(UUID.fromString("00000000-0000-0000-0000-000000000000"), new Config(null, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH), new State<>(buttonValue, StateData.empty()), 1L, 2L);
        assertThat(res).isPresent();
    }

    @Test
    public void shouldKeepExistingButtonMessage() {
        assertThat(underTest.shouldKeepExistingButtonMessage(Mockito.mock(ButtonEventAdaptor.class))).isTrue();
    }

    @Test
    public void getAnswer() {
        assertThat(underTest.getAnswer(null, null, 0L, 0L)).isEmpty();
    }

    @Test
    public void matchingComponentCustomId() {
        boolean res = underTest.matchingComponentCustomId("welcome,fate");

        assertThat(res).isTrue();
    }

    @Test
    public void matchingComponentCustomId_notMatch() {
        boolean res = underTest.matchingComponentCustomId("welcome2,fate");

        assertThat(res).isFalse();
    }

    @Test
    void matchingComponentCustomId_match() {
        assertThat(underTest.matchingComponentCustomId("welcomefate")).isTrue();
    }

    @Test
    void matchingComponentCustomId_noMatch() {
        assertThat(underTest.matchingComponentCustomId("welcome2fate")).isFalse();
    }

    @Test
    public void getCommandDefinition() {
        expect.toMatchSnapshot(underTest.getCommandDefinition());
    }

    @Test
    public void getId() {
        expect.toMatchSnapshot(underTest.getCommandId());
    }

    @Test
    void checkPersistence() {
        PersistenceManager persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
        UUID configUUID = UUID.fromString("00000000-0000-0000-0000-000000000000");
        Config config = new Config(123L, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.GERMAN);
        Optional<MessageConfigDTO> toSave = underTest.createMessageConfig(configUUID, 1L, 2L, config);
        assertThat(toSave).isPresent();

        persistenceManager.saveMessageConfig(toSave.get());
        MessageConfigDTO loaded = persistenceManager.getMessageConfig(configUUID).orElseThrow();

        assertThat(toSave.get()).isEqualTo(loaded);
        ConfigAndState<Config, StateData> configAndState = underTest.deserializeAndUpdateState(loaded, "3");

        assertThat(configAndState.getConfig()).isEqualTo(config);
        assertThat(configAndState.getConfigUUID()).isEqualTo(configUUID);
        assertThat(configAndState.getState()).isEqualTo(new State<>("3", StateData.empty()));
    }

    @Test
    void deserialization() {
        UUID configUUID = UUID.fromString("00000000-0000-0000-0000-000000000000");
        MessageConfigDTO savedData = new MessageConfigDTO(configUUID, 1L, 1660644934298L, "welcome", "Config", """
                ---
                answerTargetChannelId: 123
                answerFormatType: "full"
                configLocale: "de"
                diceStyleAndColor:
                  diceImageStyle: "none"
                  configuredDefaultColor: "none"
                """);


        ConfigAndState<Config, StateData> configAndState = underTest.deserializeAndUpdateState(savedData, "3");
        assertThat(configAndState.getConfig()).isEqualTo(new Config(123L, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.GERMAN));
        assertThat(configAndState.getConfigUUID()).isEqualTo(configUUID);
        assertThat(configAndState.getState().getData()).isEqualTo(StateData.empty());
    }

    @Test
    void configSerialization(){
        UUID configUUID = UUID.fromString("00000000-0000-0000-0000-000000000000");
        Config config = new Config(123L, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.GERMAN);
        Optional<MessageConfigDTO> toSave = underTest.createMessageConfig(configUUID, 1L, 2L, config);
        assertThat(toSave).isPresent();

        expect.toMatchSnapshot(toSave.get());
    }
}