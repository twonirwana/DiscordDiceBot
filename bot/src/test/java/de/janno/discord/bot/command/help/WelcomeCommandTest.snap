de.janno.discord.bot.command.help.WelcomeCommandTest.configSerialization=[
  {
    "configUUID" : "00000000-0000-0000-0000-000000000000",
    "guildId" : 1,
    "channelId" : 2,
    "commandId" : "welcome",
    "configClassId" : "Config",
    "config" : "---\nanswerTargetChannelId: 123\nanswerFormatType: \"full\"\nconfigLocale: \"de\"\ndiceStyleAndColor:\n  diceImageStyle: \"none\"\n  configuredDefaultColor: \"none\"\n"
  }
]


de.janno.discord.bot.command.help.WelcomeCommandTest.getCommandDefinition=[
  {
    "name" : "welcome",
    "description" : "Displays the welcome message",
    "nameLocales" : [ ],
    "descriptionLocales" : [
      {
        "locale" : "de",
        "description" : "Erzeuge eine Wegkommensnachricht"
      },
      {
        "locale" : "pt_BR",
        "description" : "Exibe a mensagem de boas vindas"
      }
    ],
    "options" : [
      {
        "type" : "SUB_COMMAND",
        "name" : "start",
        "nameLocales" : [
          {
            "locale" : "pt_BR",
            "name" : "iniciar"
          }
        ],
        "description" : "Displays the welcome message",
        "descriptionLocales" : [
          {
            "locale" : "de",
            "description" : "Erzeuge eine Wegkommensnachricht"
          },
          {
            "locale" : "pt_BR",
            "description" : "Exibe a mensagem de boas vindas"
          }
        ],
        "required" : false,
        "choices" : [ ],
        "options" : [
          {
            "type" : "STRING",
            "name" : "language",
            "nameLocales" : [
              {
                "locale" : "de",
                "name" : "sprache"
              },
              {
                "locale" : "pt_BR",
                "name" : "idioma"
              }
            ],
            "description" : "The language of the bot messages",
            "descriptionLocales" : [
              {
                "locale" : "de",
                "description" : "Die Sprach des Bots"
              },
              {
                "locale" : "pt_BR",
                "description" : "O idioma da mensagem do bot"
              }
            ],
            "required" : false,
            "choices" : [
              {
                "name" : "English",
                "value" : "en",
                "nameLocales" : [
                  {
                    "locale" : "de",
                    "choice" : "Englisch"
                  },
                  {
                    "locale" : "pt_BR",
                    "choice" : "Inglês"
                  }
                ]
              },
              {
                "name" : "German",
                "value" : "de",
                "nameLocales" : [
                  {
                    "locale" : "de",
                    "choice" : "Deutsch"
                  },
                  {
                    "locale" : "pt_BR",
                    "choice" : "Alemão"
                  }
                ]
              },
              {
                "name" : "Portuguese (Brazil)",
                "value" : "pt_BR",
                "nameLocales" : [
                  {
                    "locale" : "de",
                    "choice" : "Brasilianisches Portugiesisch"
                  },
                  {
                    "locale" : "pt_BR",
                    "choice" : "Português Brasileiro"
                  }
                ]
              }
            ],
            "options" : [ ],
            "autoComplete" : false
          }
        ],
        "autoComplete" : false
      },
      {
        "type" : "SUB_COMMAND",
        "name" : "help",
        "nameLocales" : [
          {
            "locale" : "de",
            "name" : "hilfe"
          },
          {
            "locale" : "pt_BR",
            "name" : "ajuda"
          }
        ],
        "description" : "Get help for /welcome",
        "descriptionLocales" : [
          {
            "locale" : "de",
            "description" : "Hilfe für /welcome"
          },
          {
            "locale" : "pt_BR",
            "description" : "Consiga ajuda para /welcome"
          }
        ],
        "required" : false,
        "choices" : [ ],
        "options" : [ ],
        "autoComplete" : false
      }
    ]
  }
]


de.janno.discord.bot.command.help.WelcomeCommandTest.getId=[
  "welcome"
]